backupccl: reduce over-splitting during RESTORE

dt opened the pull request:
When constructing the spans to restore, we can produce fewer, wider spans
that each contain more files to avoid splitting as much.

Previously we split at the bounds of every file in the base backup, then
added incremental layer files to any span they overlapped. This changes
that slightly to keep appending files that would create a new span at the
right-hand-side of the span covering to instead extend current rightmost
entry in the covering if the total size of what has been added to it is
still below our target size per span (384mb).

Release justification: bug fix.

Release note (bug fix): reduce the amount that RESTORE over-splits.

dt committed:
backupccl: reduce over-splitting during RESTORE

When constructing the spans to restore, we can produce fewer, wider spans
that each contain more files to avoid splitting as much.

Previously we split at the bounds of every file in the base backup, then
added incremental layer files to any span they overlapped. This changes
that slightly to keep appending files that would create a new span at the
right-hand-side of the span covering to instead extend current rightmost
entry in the covering if the total size of what has been added to it is
still below our target size per span (384mb).

Release justification: bug fix.

Release note (bug fix): reduce the amount that RESTORE over-splits.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86496)
<!-- Reviewable:end -->

dt commented:
Here's a of the same backup we use in the restore2tb backup -- with is roughly 660GB data size -- being restored with and without (an earlier version of) this patch. Note the axis labels:

<img width="982" alt="Screen Shot 2022-08-19 at 6 34 11 PM" src="https://user-images.githubusercontent.com/15615/185715646-b2719f1b-f5aa-4ee5-adad-26f1d746df69.png">

<img width="986" alt="Screen Shot 2022-08-19 at 6 34 26 PM" src="https://user-images.githubusercontent.com/15615/185715647-98b61fa7-066e-45a5-ba7c-cc8d8203dc58.png">

I think I was using a slightly different target size and lower guess sizes for that run (I believe they were 448 and 8mb, but I could be wrong). I did notice some rebalancing traffic on the merged spans cluster however, indicating we were not splitting quite enough to avoid any post-ingest rebalance:
<img width="991" alt="Screen Shot 2022-08-19 at 6 46 58 PM" src="https://user-images.githubusercontent.com/15615/185716534-046746fb-d29e-47a3-88fa-3b175c0ff18a.png">
I suspect we'd benefit from enabling as-we-fill splits in the sst batcher to avoid these post-ingest rebalances. The amount however isn't very concerning and didn't have a noticeable impact on top-line restore time:
41:00 on the control master, and 40:17 on the experimental merging build. The merging build's rebalancing did mean that final disk usage on that cluster was 2.1% higher than the master one where everything went straight to L6 and never moved.

msbutler commented:
wow, beat me to it! I've got another impl that seems slightly more involved [here](https://github.com/msbutler/cockroach/commit/1bf0004645e6e245b388e59c6d4cddf72183480a). I'll plan to dig a little deeper here next week.

msbutler commented:
will also consider how this jives with @itsbilal [pr](https://github.com/cockroachdb/pebble/pull/1905)

dt commented:
@msbutler took a peek at your version, but I _think_ prefer just widening the rightmost cover span as we go vs needing to mess with the the covering after. I think I'd suggest you take this PR over instead but up to to. I think the biggest thing I'd want to do here to merge this is add another test case where the size isn't zero, and maybe make it a cluster setting. I'd also like to enable as-we-go-splits in the batcher but I think my roachtest run shows that that isn't a strict requirement.

On the level merging iter: I'm not really sure this changes things? This is putting more `Files` into each span, but there was already no real guarantees about sstables -- not files -- not overlapping in a level (indeed they overlap all the time). The only difference is previously in the _first level_ we'd never assign multiple Files to the same span, but now we will, but we can and would often assign multiple files to the same Span in inc levels, so there was already the potential for overlapping sstables (since even thought the files are non-overlapping, they can be in different sstables which _are_ overlapping), right?

msbutler commented:
ah rats, Bilal's PR cannot get applied to RESTORE. I was conflating `files` and `SSTable` in `makeSimpleImportSpans`

msbutler commented:
The high level difference between David's implementation and mine is:
- my implementation expands the spans `after` the `restoreSpanEntry` has formed, which provides an upper bound on the amount of work in a span. The upper bound prevents mid/post RESTORE range splits (full of data).
- David's expands the spans _during_ `restoreSpanEntry` formation and only when evaluating the 0th layer (the full backup spans). This impl cannot guarantee an upper bound of work on a given span. So, if incremental backups add significantly more work to a span, this implementation could cause mid/post restore range splits, which are _much_ more expensive than the mid/post restore range merges that we're currently dealing with. 

Given all this, I think it makes sense to go with david's lighter weight implementation, as long as we also implement range splitting in the batcher.

msbutler commented:
@dt per weekly standup, I'm handing this back to you.

dt committed:
backupccl: enable splitting full ranges during restore

Previously sstBatcher had a boolean to control whether it would split
and then scatter a range it was filling when it could determine that it
was about to overfill it i.e. if had previously reported its remaining
capacity to be less than what the batcher was about to add. RESTORE set
this boolean to disable this splitting and scattering as it pre-split and
pre-scattered all of its ranges and thus did not expect the batcher to
need to do any as-it-fills splitting and scattering.

However blanket disabling this makes little sense: if the batcher knows
that it is about to overfill a range, regardless of whether or not that
range was supposed to be pre-split to avoid this, it always makes sense
to split it before adding to it, rather than overfill it.

Thus this change makes as-we-fill splits unconditional, and moves the
boolean to be more narrowly scoped to just disabling the scatters, as
RESTORE still pre-scatters its pre-chunked restore spans. This paves the
way for RESTORE to make and pre-split larger chunks, without worrying if
doing so will cause overfilling of ranges.

Release note: none.

Release justification: bug fix.

dt commented:
Added another commit to enable as-we-fill splits of full ranges, replacing the disabling of batcher's splits during restore more narrow, to be just a disabling of scatters.
Results look promising (note the axis labels i.e the magnitude of these are not comparable):

master, all pre-split and very overspilt (35min, 21602 ranges)
<img width="970" alt="Screen Shot 2022-08-30 at 9 11 33 PM" src="https://user-images.githubusercontent.com/15615/187571190-e3d6be35-af6f-4eb0-b1b3-b9259bca4218.png">

just bigger spans, no as-we-go splits (32min, 1976 ranges):
<img width="972" alt="Screen Shot 2022-08-30 at 9 11 15 PM" src="https://user-images.githubusercontent.com/15615/187571270-1a60053a-f46c-4b47-961a-ef83a03599a2.png">

bigger spans plus enabling as-we-go-splits (29min, 1953 ranges):
<img width="969" alt="Screen Shot 2022-08-30 at 9 11 48 PM" src="https://user-images.githubusercontent.com/15615/187571304-bf2cb21c-bde1-44d5-8038-730480543bf8.png">

dt commented:
> how late we are in the release, I don't think any of this should be enabled by default.

I'm tempted to enable it, and potentially even backport too, to fix the bad behavior of over-splitting, at least to 22.1.

shermanCRL commented:
> > how late we are in the release, I don't think any of this should be enabled by default.

> I'm tempted to enable it, and potentially even backport too, to fix the bad behavior of over-splitting, at least to 22.1. I'm not _that_ scared of it, since we're just changing how we divide the spans, not what we do with them or how we do it really.

You can make the call, I’m ok with it if the tests are good (including the cluster testing we plan to do).

msbutler reviewed (commented):
I have two high level worries about this diff (that are perhaps off base): 
First, this diff makes it easier for one node to end up processing significantly more data, relative to other nodes in the flow. This has a couple risky consequences:
1. the restore runtime can get bottlenecked by one node
2. kv may feel compelled to move full ranges from the hot node, which is probably more expensive than merging small ranges

Here's one way this could happen, even with range splitting enabled:
- suppose the incremental backups add significantly more data to a single `restoreSpanEntry` that `makeSimpleImportSpans` has grown, as the full backup contained relatively less data in this span. Also suppose that before this diff, the data from the incremental backups would have been divided into several `restoreSpanEntry`'s that also got divided over several import span chunks.
- before this diff, the incremental backup data was divided across chunks, and therefore, work was distributed across the cluster.
- with this diff, it is more likely that al the incremental backup data will end up in fewer chunks, and therefore get processed by fewer nodes. 

Second, given that this diff does not change our checkpointing procedure (something we said should happen after this work), the larger spans this work creates makes checkpointing more imprecise, which could make restore job pausing/resuming more expensive. (i.e. https://github.com/cockroachdb/cockroach/issues/81116, though disallowShadowing [may have been disabled](https://github.com/cockroachdb/cockroach/issues/81116#issuecomment-1135128433)? ).

dt commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -87,6 +97,11 @@ func makeSimpleImportSpans(
 						cover = append(cover, makeEntry(span.Key, sp.EndKey, fileSpec))
+						if f.EntryCounts.DataSize > 0 {
+							curCovGrowSize = f.EntryCounts.DataSize
+						} else {
```
old backups don't set EntryCounts.

dt commented:
> suppose the incremental backups add significantly more data to a single restoreSpanEntry that makeSimpleImportSpans has grown, as the full backup contained relatively less data in this span

I thought a little about this initially, but came away thinking this is true _anyway_, even without merging base spans at all, so we shouldn't let it stop us from fixing the issue at hand. If we think about when we might hit this case -- where the base backup has very little data in some span but then later layers have much more in that span, that is going to be most common when you create a table, back it up, then load data into it, i.e. many ranges of data, then backing up incrementally; all those big, full ranges of the inc backup all fall within the one span in the inc backup. This however is all true without merging -- it's a single span per table when it is empty -- so the only difference with merging is magnitude, but my thesis here is that magnitudes that are unfortunate are reachable anyway.

If we decide that we care about this -- I'm not sure we really do right now -- then I think we'd want to do something more general, that post-processes the covering if any one chunk is more than X% (probably 1/nodes*2 or something), but I don't really see it as coupled to merging small base spans.

> does not change our checkpointing procedure (something we said should happen after this work)

I don't think we should make extra spans here just for the sake of progress checkpoints since we already know that that does not actually make progress lose that much less; our losses are already due to out-of-order span completion, not individual span sizes, and even post-merging our spans sizes here are still, in absolute numbers, generally small. I don't really find the progress effects that compelling: this was always intended to have ~range sized chunks, is currently dramatically undershooting that, and 8mb or 64mb or 364mb or even 3gb are all tiny numbers in TB restores, where we really care about progress. In a tiny restore where the difference here is actually a real % of the total size, then the total size is small enough to not really care about progress much anyway since it is small anyway (also there's a minimum time between progress updates too).

msbutler commented:
Cool, I'm no longer worried about merging creating an uneven distribution of work.

Sorry, I should have been more clear on why I worry about checkpointing. Like you, I don't care that much about the additional lost work, given that the spans are larger. I worry that the larger spans will cause Restore to slow to crawl more frequently due to CheckForSSTCollisions. In [22.1](https://github.com/cockroachdb/cockroach/pull/82357) and 21.2, we set `disallowShadowing` to false. Perhaps in this PR or after the branch cut, we should set disallowShadowing to false in 22.2 as well.

msbutler commented:
> The lost-work and the disallowShadowing shadowing overlaps are one and the same though, aren't they?
Yes, both are caused by our naive ordering-sensitive checkpointing strategy, which, as you say, causes us to reingest a bunch of chunks again. My point is: if the avg span entry size is significantly bigger and the number of span entries retried remains constant (not sure if this is correct), then this diff means we're more susceptible to the check for key collisions problem. Regardless, I think we should set `disallowShadowing` == false for 22.2, as we have not yet updated our checkpointing scheme, which means my point is moot.

dt commented on pkg/ccl/backupccl/restore_span_covering_test.go:
```diff
@@ -119,6 +119,8 @@ func checkRestoreCovering(
+
 func TestRestoreEntryCoverExample(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
```
Yep, done!

dt commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -30,6 +30,8 @@ func (ie intervalSpan) Range() interval.Range {
 	return interval.Range{Start: []byte(ie.Key), End: []byte(ie.EndKey)}
 }
 
```
done

msbutler commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -78,20 +99,55 @@ func makeSimpleImportSpans(
 
 		for layer := range backups {
 			covPos := spanCoverStart
+			var curCovGrowSize int64
```
maybe add def `curGovGrowSize`? I think it's: curCovGrowSize is the number of bytes added to the right most cover since its creation.

msbutler commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -78,20 +99,55 @@ func makeSimpleImportSpans(
+								if l := len(cover[i].Files) - 1; l < 0 || cover[i].Files[l] != fileSpec {
+									cover[i].Files = append(cover[i].Files, fileSpec)
+									if i == l {
+										curCovGrowSize += sz
```
I might be missing something here. why do we need grow `curCovGrowSize` here for when `i==l`? 

I thought this loop is growing "older" covers (i.e. not the current, rightmost cover), potentially beyond the targetSize limit.  I'm under the impression that curCovGrowSize tracks the right most cover's size, so I don't follow why it gets incremented here.

Perhaps this logic could use another comment. It's quite hard to follow all the indexing!

dt replied:
`i == l` is the rightmost

msbutler replied:
didn't we already grow the right most span above?

msbutler replied:
aha! This makes sense now. thanks for clarifying!

msbutler commented:
The non test code looks good to me. Last question: how confident are you that the test cases you put in exercise every nook and cranny of  the new makeSimpleImport spans? Before backporting, I wonder if we ought to write a few more specific test cases to ensure this. `makeSimpleImportSpans` is no longer so simple :)

dt commented:
> how confident are you that the test cases you put in exercise every nook and cranny of the new makeSimpleImport spans

The random test does a good job of ensuring correctness (it caught a bug here). It doesn't do much in terms of proving we merged into any specific number of spans for perf, but it proves we have the right files in the right spans I think, which makes me feel better about this.

dt commented:
TFTR!

bors r+

dt commented:
I tagged it for backport; we can see how long we want to bake on 22.2 before we merge it

dankinder commented:
@dt I have been waiting eagerly for this change as I'm still blocked on restoring, per https://github.com/cockroachdb/cockroach/issues/86470 so thanks for cranking it out.

I'm currently on V22.1.5. I'll either need to wait until this gets backported to 22.1, or deploy the next alpha of 22.2, or build it myself. If you have any suggestion on which is best or the timeline of when these will happen I'd appreciate it.
